msprime/mutations.py

```diff
@@ -199,6 +199,301 @@ def __init__(self, state_independent=False):
         super().__init__(alleles, root_distribution, transition_matrix)
 
 
+def general_microsat_rate_matrix(s, u, v, p, m, lo, hi):
+    """
+    Construct the general microsat rate matrix associated
+    with the Sainudiin et al. (2004) parameterization.
+    While the original paper implements a continuous time markov chain,
+    here we implement the jump process rate matrix equivalent for consistency
+    with the :class:`.MatrixMutationModel`.
+
+    In particular if :math:`q_{ij}` is an element of the originally defined
+    transition matrix, we define :math:`M = \\max_i \\left( \\sum_j q_{ij} \\right)`.
+    Then, as a :class:`.MatrixMutationModel` this model has the transition
+    matrix :math:`P_{ij} = q_{ij} / M`, and :math:`P_{ii} = 1 - \\sum_{j\\neq i} P_{ij}`.
+
+    :param float s: strength of length dependence on mutation rate
+     (i.e. s=0 uniform rate)
+    :param float u: constant bias parameter; U ~ (0, 1)
+    :param float v: linear bias parameter; v ~ (-inf, inf)
+    :param float p: probability of a single step mutation
+    :param float m: success prob of truncated gamma distribution
+    :param int lo: lower bound on repeat number
+    :param int hi: upper bound on repeat number
+
+    :return: rate matrix
+
+    """
+
+    # Helper functions for the Sainudiin et al. rate matrix
+    def alpha_microsat(i, u, v, lo):
+        return max(0, min(1, u - (v * (i - lo))))
+
+    def beta_microsat(i, s, lo):
+        """
+        beta helper function from Sainudiin et al. (2004)
+        represents the mutation rate from allele i
+        :param float s: strength of length dependence on mutation rate
+        (i.e. s=0 uniform rate)
+        :param int lo: lower bound
+        :param int hi: upper bound
+        """
+        return 1 + ((i - lo) * s)
```

Comment:
Bad things happen if `s < -1/(hi - lo)`; we should probably check for and error in this case? (gee, if `s = -1/(hi-lo)` then `hi` is an absorbing state, and the eigenvalue computation probably fails?)

Reply:
Ah in Sainudiin they say that `s > - 1/(hi - lo + 1)`.

Reply:
okay i'll add this as a specific test in `MicrosatMutationModel`